### PR TITLE
Add Vote Phasing

### LINKED ISSUE #434: Implement Vote and Round phasing in the Voting contract

### PR DESCRIPTION
**Pull request by mrice32**
This PR adds phasing to the voting contract. It also adds a new library for vote phasing.

Fixes #434.

### CODE REVIEW

File: core/contracts/Voting.sol [hunk 1]
```diff
@@ -113,55 +122,66 @@ contract Voting is Testable {
...
         uint priceResolutionRound = _getPriceResolution(identifier, time).lastVotingRound;
-        uint secondsInWeek = 60*60*24*7;
 
         if (priceResolutionRound == 0) {
             // Price has never been requested.
+
+            // Must ensure the round is updated here so the requested price will be voted on in the next commit cycle.
+            // TODO: this may be an expensive operation - may make sense to have a public updateRound() method to handle
+            // extreme cases.
+            _updateRound(blockTime);
```
Comment (ptare) [inline core/contracts/Voting.sol]: Maybe worth adding a comment here explaining that we don't update the round unless we have to, because we want to push the burden of doing the update onto voters, not financial contracts (indirectly token sponsors). [is that what we want to do?]

File: core/test/Voting.js [hunk 2]
```diff
@@ -3,30 +3,76 @@ const { didContractThrow } = require("../../common/SolidityTestUtils.js");
...
+    await moveToNextRound();
+
+    // Rounds should start with Commit.
+    assert.equal((await voting.getVotePhase()).toString(), "0");
+
+    // Shift of one phase should be Reveal.
+    await moveToNextPhase();
+    assert.equal((await voting.getVotePhase()).toString(), "1");
+
+    // A seconds shift should go back to commit.
```
Comment (ptare) [inline core/test/Voting.js]: ultra-nit: second not seconds.

File: core/contracts/VoteTiming.sol [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        data.phaseLength = phaseLength;
+        data.roundId = 1;
+        data.roundStartTime = 0;
+    }
+
+    function getLastUpdatedRoundId(Data storage data) internal view returns (uint) {
+        return data.roundId;
+    }
+
+    function shouldUpdateRoundId(Data storage data, uint currentTime) internal view returns (bool) {
```
Comment (ptare) [inline core/contracts/VoteTiming.sol]: So this function is in plain English:  "has the stored round ended"
and `updateRoundId` is: "start new round"

Reminds me a bunch of my struggle with understanding "previous", "current", and "next" in the TokenizedDerivative :)

**Comment by mrice32** [comment]
@ptare @rawfalafel I just added a wrapper contract and tests for the library. It also included a small refactor of the library to simplify it. Since it doesn't look like any of the comments centered around that, it shouldn't impact any of y'all's previous review.

File: core/contracts/VoteTiming.sol [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+    function updateRoundId(Data storage data, uint currentTime) internal {
+        data.roundId = computeCurrentRoundId(data, currentTime);
+        data.roundStartTime = _computeCurrentRoundStartTime(data, currentTime);
+    }
+
+    function computeCurrentRoundId(Data storage data, uint currentTime) internal view returns (uint roundId) {
+        roundId = data.roundId;
+        if (shouldUpdateRoundId(data, currentTime)) {
+            roundId = roundId.add(1);
```
[the remark this answers is not in the document]
Comment (mrice32) [reply core/contracts/VoteTiming.sol]: The idea is that rounds should only ever increment by `1`. So if there is no vote for 6 months, the vote round still increments by `1` when a new one occurs. Otherwise, we could statelessly compute the round ID from the time.

File: core/contracts/Voting.sol [hunk 5]
```diff
@@ -219,4 +308,39 @@ contract Voting is Testable {
         bytes memory encodedArgs = abi.encode(identifier, time);
         return priceResolutions[encodedArgs];
     }
+
+    function _updateRound(uint blockTime) private {
+        if (voteTiming.shouldUpdateRoundId(blockTime)) {
```
[the remark this answers is not in the document]
Comment (mrice32) [reply core/contracts/Voting.sol]: Haha, you and @ptare both flagged this. Good point. Will do.

[the remark this answers is not in the document]
Comment (mrice32) [reply core/contracts/Voting.sol]: I love that y'all both immediately commented on this - I'll update.

File: core/contracts/VoteTiming.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (rawfalafel) [reply core/contracts/VoteTiming.sol]: Doesn't that assumption contradict `computeEstimatedRoundEndTime` ? Looks like that function assumes no skips to get the time.

File: core/contracts/Voting.sol [hunk 6]
```diff
@@ -170,7 +190,86 @@ contract Voting is Testable {
...
+    function initializeOnce(uint phaseLength) public {
+        require(!initialized, "Only the constructor should call this method");
+        initialized = true;
+        voteTiming.init(phaseLength);
+    }
+
+    function _getRolloverPriceRequests(uint roundId)
+        private
+        view
+        returns (PriceRequest[] memory rolloverPriceRequests)
```
[the remark this answers is not in the document]
Comment (mrice32) [reply core/contracts/Voting.sol]: Done.

File: core/contracts/Voting.sol [hunk 7]
```diff
@@ -113,55 +122,66 @@ contract Voting is Testable {
...
+            _addPriceRequestToRound(nextRoundId, PriceRequest(identifier, time));
+
+            // Estimate the end of next round and return the time.
+            return voteTiming.computeEstimatedRoundEndTime(nextRoundId);
+        } else if (priceResolutionRound >= currentRoundId) {
+            // Price is already slated to be resolved.
+            
+            return voteTiming.computeEstimatedRoundEndTime(priceResolutionRound);
         } else {
             // Price has been resolved.
```
[the remark this answers is not in the document]
Comment (mrice32) [reply core/contracts/Voting.sol]: Good point, I just moved the call up for now because keeping it in the if statement seems like it would make the code more confusing as you mentioned earlier today.

File: core/contracts/Voting.sol [hunk 8]
```diff
@@ -113,55 +122,66 @@ contract Voting is Testable {
...
-            // TODO: replace this with the end of the next round once phasing has been implemented.
-            return getCurrentTime().add(secondsInWeek.mul(2));
-        } else if (priceResolutionRound == currentRoundNumber) {
-            // Price is currently being resolved.
-            // TODO: replace this with the end of the current round once phasing has been implemented.
-            return getCurrentTime().add(secondsInWeek);
+            _addPriceRequestToRound(nextRoundId, PriceRequest(identifier, time));
+
+            // Estimate the end of next round and return the time.
+            return voteTiming.computeEstimatedRoundEndTime(nextRoundId);
```
[the remark this answers is not in the document]
Comment (mrice32) [reply core/contracts/Voting.sol]: See the comment in our internal doc.

File: core/contracts/Voting.sol [hunk 9]
```diff
@@ -113,55 +122,66 @@ contract Voting is Testable {
...
         if (priceResolutionRound == 0) {
             // Price has never been requested.
+
+            // Must ensure the round is updated here so the requested price will be voted on in the next commit cycle.
+            // TODO: this may be an expensive operation - may make sense to have a public updateRound() method to handle
+            // extreme cases.
+            _updateRound(blockTime);
+
+            // Price requests always go in the next round, so add 1 to the computed current round.
+            uint nextRoundId = currentRoundId.add(1);
```
[the remark this answers is not in the document]
Comment (mrice32) [reply core/contracts/Voting.sol]: Added a comment to the section that you linked.

File: core/contracts/Voting.sol (hunk 1 above)
Comment (mrice32) [reply core/contracts/Voting.sol]: Added a comment and moved the call up.

File: core/contracts/VoteTiming.sol [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        }
+    }
+
+    function computeCurrentPhase(Data storage data, uint currentTime) internal view returns (Phase) {
+        // This employs some hacky casting. We could make this an if-statement if we're worried about type safety.
+        return Phase(currentTime.div(data.phaseLength).mod(NUM_PHASES));
+    }
+
+    function computeEstimatedRoundEndTime(Data storage data, uint roundId) internal view returns (uint) {
```
[the remark this answers is not in the document]
Comment (mrice32) [reply core/contracts/VoteTiming.sol]: I'll add it to the docstring `@dev`.

File: core/contracts/VoteTiming.sol (hunk 3 above)
Comment (mrice32) [reply core/contracts/VoteTiming.sol]: Do you have a suggestion for a different set of names? I don't really like the ones I went with, so suggestions are welcome :)

File: core/test/Voting.js [hunk 11]
```diff
@@ -3,30 +3,76 @@ const { didContractThrow } = require("../../common/SolidityTestUtils.js");
...
+  before(async function() {
+    voting = await Voting.deployed();
+  });
+
+  it("Vote phasing", async function() {
+    // Reset the rounds.
+    await moveToNextRound();
+
+    // Rounds should start with Commit.
+    assert.equal((await voting.getVotePhase()).toString(), "0");
```
[the remark this answers is not in the document]
Comment (mrice32) [reply core/test/Voting.js]: Done.

File: core/test/Voting.js (hunk 2 above)
Comment (mrice32) [reply core/test/Voting.js]: Done.

**Comment by mrice32** [comment]
Added docstrings.
